pkg/pmem-csi-operator/controller/deployment/deployment_controller.go

```diff
@@ -47,25 +53,128 @@ func Add(mgr manager.Manager, opts pmemcontroller.ControllerOptions) error {
+			changes := old.Compare(new)
+			klog.Infof("CR changes: %v", changes)
+			return len(changes) != 0
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			r.reconcileMutex.Lock()
+			defer r.reconcileMutex.Unlock()
+			klog.Infof("DELETED: %T/%s", e.Object, e.Meta.GetName())
+			// Deployment CR deleted, remove it's reference from cache.
+			// Objects owned by it are automatically garbage collected.
+			r.deleteDeployment(e.Meta.GetName())
+			// We already handled the event here,
+			// so no more further reconcile required.
+			return false
+		},
+	}
+
 	// Watch for changes to primary resource Deployment
-	err = c.Watch(&source.Kind{Type: &pmemcsiv1alpha1.Deployment{}}, &handler.EnqueueRequestForObject{})
-	if err != nil {
+	if err := c.Watch(&source.Kind{Type: &pmemcsiv1alpha1.Deployment{}}, &handler.EnqueueRequestForObject{}, p); err != nil {
 		klog.Errorf("Deployment.Add: watch error: %v", err)
 		return err
 	}
 
+	// Predicate functions for sub-object changes
+	// We do not want to pollute deployment reconcile loop by sending
+	// sub-object changes, instead we provide a dedicated handler.
+	// So all these event handlers returns 'false' so that the event
+	// is not propagated further.
+	// One exception is: If we fail to handle here, then we pass this
+	// event to reconcile loop, where it should recognize these requests
+	// and just requeue. Expecting that the failure is retried.
+	eventFunc := func(meta metav1.Object, obj apiruntime.Object) bool {
+		// Get the owned deployment
+		d, err := r.getDeploymentFor(meta)
+		if err != nil {
+			klog.Infof("%T/%s is not owned by any deployment", obj, meta.GetName())
+			// The owner might have deleted already
+			// we can safely ignore this event
+			return false
+		}
+		r.reconcileMutex.Lock()
+		defer r.reconcileMutex.Unlock()
+		if err := d.HandleEvent(meta, obj, r); err != nil {
+			klog.Warningf("Error occurred while handling the event: %v, requeuing the event", err)
+			return true
+		}
+		return false
+	}
+	sop := predicate.Funcs{
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			klog.Infof("DELETED: %T/%s", e.Object, e.Meta.GetName())
+			return eventFunc(e.Meta, e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.MetaNew.GetDeletionTimestamp() != nil {
+				// We can handle this in delete handler
+				return false
+			}
+			klog.Infof("UPDATED: %T/%s", e.ObjectOld, e.MetaOld.GetName())
+			return eventFunc(e.MetaOld, e.ObjectOld)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Do not handle sub-object create events
+			return false
+		},
+	}
+
+	// All possible object types for a Deployment CR
+	// that would requires recdeploying on change
+	subresources := []runtime.Object{
+		&rbacv1.ClusterRole{},
+		&rbacv1.ClusterRoleBinding{},
+		&storagev1beta1.CSIDriver{},
+		&appsv1.DaemonSet{},
+		&rbacv1.Role{},
+		&rbacv1.RoleBinding{},
+		&corev1.Secret{},
+		&corev1.Service{},
+		&corev1.ServiceAccount{},
+		&appsv1.StatefulSet{},
+	}
```

Comment:
Don't we already have such a list? If the redundancy cannot be avoided, then at least add comments in both places about keeping the lists in sync.

Reply:
This list is different from the `AllObjectTypes` list defined in `controller_driver.go`. As suggested added a note in the comment.

Reply:
> I'm not convinced that we need a second list.

The AllObjectTypes in `controller_driver.go` might also contain objects that were obsolete now, whereas this watch list is supposed to have only active objects requires reconciling.

Reply:
> See [avalluri#3](https://github.com/avalluri/pmem-CSI/pull/3)

Splitting `AllObjectTypes` list into current + obsolete seems works fine. I took the patch. Thanks.